undefined reference to `xlnt::uri::is_relative() const'

With: 
```cpp
xlnt::hyperlink cell_hyperlink = cell.hyperlink();
xlnt::uri cell_hyperlink_relationship_source = cell_hyperlink_relationship.source();
if (cell_hyperlink_relationship_source.is_relative())
{
    std::cout << "cell_hyperlink_relationship_source is relative" << std::endl;
}
```

I get : 
```
undefined reference to `xlnt::uri::is_relative() const'
```

But:

https://xlnt-community.github.io/xlnt/classxlnt_1_1uri.html

![Image](https://github.com/user-attachments/assets/bbf5aec0-a9ed-4dfb-aa40-d5536d0b7764)

--- a/include/xlnt/packaging/uri.hpp
+++ b/include/xlnt/packaging/uri.hpp
@@ -81,6 +81,11 @@ public:
     std::string authority() const;
 
     /// <summary>
+    /// Returns true if this URI explicitly contains an authority component.
+    /// </summary>
+    bool has_authority() const;
+
+    /// <summary>
     /// Returns true if an authentication section is specified for this URI.
     /// </summary>
     bool has_authentication() const;
@@ -180,6 +185,12 @@ private:
     /// True if this URI is absolute.
     /// </summary>
     bool absolute_ = false;
+
+    /// <summary>
+    /// True if this URI explicitly contains an authority component ("//").
+    /// RFC 3986 distinguishes authority presence from whether host/auth/port exist.
+    /// </summary>
+    bool has_authority_ = false;
 
     /// <summary>
     /// True if this URI has an authentication section.

--- a/include/xlnt/packaging/uri.hpp
+++ b/include/xlnt/packaging/uri.hpp
@@ -193,6 +193,11 @@ private:
     bool has_authority_ = false;
 
     /// <summary>
+    /// True if the host is an IPv6 address and needs bracket wrapping in authority().
+    /// </summary>
+    bool ipv6_ = false;
+
+    /// <summary>
     /// True if this URI has an authentication section.
     /// </summary>
     bool has_authentication_ = false;

--- a/source/cell/cell.cpp
+++ b/source/cell/cell.cpp
@@ -451,7 +451,7 @@ void cell::hyperlink(const std::string &url, const std::string &display)
     // check for existing relationships
     auto relationships = manifest.relationships(ws.path(), relationship_type::hyperlink);
     auto relation = std::find_if(relationships.cbegin(), relationships.cend(),
-        [&url](const xlnt::relationship &rel) { return rel.target().path().string() == url; });
+    [&url](const xlnt::relationship &rel) { return rel.target().to_string() == url; });
     if (relation != relationships.end())
     {
         d_->hyperlink_.get().relationship = *relation;

--- a/source/detail/serialization/xlsx_consumer.cpp
+++ b/source/detail/serialization/xlsx_consumer.cpp
@@ -1070,7 +1070,7 @@ worksheet xlsx_consumer::read_worksheet_end(const std::string &rel_id)
 
                     if (hyperlink_rel != hyperlinks.end())
                     {
-                        auto url = hyperlink_rel->target().path().string();
+                        auto url = hyperlink_rel->target().to_string();
 
                         if (cell.has_value())
                         {

--- a/source/detail/serialization/xlsx_producer.cpp
+++ b/source/detail/serialization/xlsx_producer.cpp
@@ -3520,7 +3520,7 @@ void xlsx_producer::write_relationships(const std::vector<xlnt::relationship> &r
 
         write_attribute("Id", relationship.id());
         write_attribute("Type", relationship.type());
-        write_attribute("Target", relationship.target().path().string());
+        write_attribute("Target", relationship.target().to_string());
 
         if (relationship.target_mode() == xlnt::target_mode::external)
         {

--- a/source/packaging/uri.cpp
+++ b/source/packaging/uri.cpp
@@ -1,29 +1,781 @@
+// Copyright (c) 2014-2022 Thomas Fussell
+// Copyright (c) 2024-2026 xlnt-community
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE
+//
+// @license: http://www.opensource.org/licenses/mit-license.php
+// @author: see AUTHORS file
+
 #include <xlnt/packaging/uri.hpp>
 
-namespace xlnt {
+#include <algorithm>
+#include <cctype>
+#include <limits>
+#include <string>
+#include <vector>
 
-uri::uri()
+namespace xlnt
 {
+namespace
+{
+bool starts_with(const std::string &s, const char *prefix)
+{
+    const std::size_t n = std::char_traits<char>::length(prefix);
+    return s.size() >= n && s.compare(0, n, prefix) == 0;
 }
+
+std::string lower_copy(std::string s)
+{
+    for (char &c : s)
+    {
+        c = static_cast<char>(std::tolower(static_cast<unsigned char>(c)));
+    }
+    return s;
+}
+
+bool is_scheme_char(char c)
+{
+    return std::isalpha(static_cast<unsigned char>(c)) ||
+           std::isdigit(static_cast<unsigned char>(c)) ||
+           c == '+' || c == '-' || c == '.';
+}
+
+bool looks_like_scheme(const std::string& s)
+{
+    if (s.empty() || !std::isalpha(static_cast<unsigned char>(s[0])))
+        return false;
+
+    for (std::size_t i = 1; i < s.size(); ++i)
+    {
+        char c = s[i];
+        if (c == ':') return true;
+        if (!is_scheme_char(c)) return false;
+    }
+    return false;
+}
+
+bool parse_port(const std::string &s, std::size_t &port)
+{
+    if (s.empty())
+    {
+        return false;
+    }
+
+    std::size_t value = 0;
+    for (char c : s)
+    {
+        if (!std::isdigit(static_cast<unsigned char>(c)))
+        {
+            return false;
+        }
+
+        const std::size_t digit = static_cast<std::size_t>(c - '0');
+        if (value > (std::numeric_limits<std::size_t>::max() - digit) / 10)
+        {
+            return false;
+        }
+
+        value = value * 10 + digit;
+    }
+
+    port = value;
+    return true;
+}
+
+std::string bracket_ipv6_if_needed(const std::string &host)
+{
+    if (host.find(':') != std::string::npos &&
+        !(host.size() >= 2 && host.front() == '[' && host.back() == ']'))
+    {
+        return "[" + host + "]";
+    }
+
+    return host;
+}
+
+struct parsed_authority
+{
+    bool present = false;
+    bool authentication = false;
+    bool port_present = false;
+    std::string username;
+    std::string password;
+    std::string host;
+    std::size_t port = 0;
+};
+
+parsed_authority parse_authority(const std::string &authority)
+{
+    parsed_authority result;
+    result.present = true;
+
+    if (authority.empty())
+    {
+        return result;
+    }
+
+    std::string hostport = authority;
+
+    const std::size_t at = authority.rfind('@');
+    if (at != std::string::npos)
+    {
+        result.authentication = true;
+        const std::string userinfo = authority.substr(0, at);
+        hostport = authority.substr(at + 1);
+
+        const std::size_t colon = userinfo.find(':');
+        if (colon != std::string::npos)
+        {
+            result.username = userinfo.substr(0, colon);
+            result.password = userinfo.substr(colon + 1);
+        }
+        else
+        {
+            result.username = userinfo;
+        }
+    }
+
+    if (!hostport.empty() && hostport.front() == '[')
+    {
+        const std::size_t close = hostport.find(']');
+        if (close != std::string::npos)
+        {
+            result.host = hostport.substr(1, close - 1);
+            if (close + 1 < hostport.size() && hostport[close + 1] == ':')
+            {
+                std::size_t port = 0;
+                if (parse_port(hostport.substr(close + 2), port))
+                {
+                    result.port_present = true;
+                    result.port = port;
+                }
+            }
+        }
+        else
+        {
+            result.host = hostport;
+        }
+    }
+    else
+    {
+        const std::size_t colon = hostport.rfind(':');
+        if (colon != std::string::npos && hostport.find(':') == colon)
+        {
+            std::size_t port = 0;
+            if (parse_port(hostport.substr(colon + 1), port))
+            {
+                result.host = hostport.substr(0, colon);
+                result.port_present = true;
+                result.port = port;
+            }
+            else
+            {
+                result.host = hostport;
+            }
+        }
+        else
+        {
+            result.host = hostport;
+        }
+    }
+
+    return result;
+}
+
+std::string remove_dot_segments(std::string path)
+{
+    std::string output;
+    while (!path.empty())
+    {
+        if (path.compare(0, 3, "../") == 0)
+        {
+            path.erase(0, 3);
+        }
+        else if (path.compare(0, 2, "./") == 0)
+        {
+            path.erase(0, 2);
+        }
+        else if (path.compare(0, 3, "/./") == 0)
+        {
+            path.replace(0, 3, "/");
+        }
+        else if (path == "/.")
+        {
+            path = "/";
+        }
+        else if (path.compare(0, 4, "/../") == 0)
+        {
+            path.replace(0, 4, "/");
+            std::size_t slash = output.find_last_of('/');
+            if (slash != std::string::npos)
+                output.erase(slash);
+            else
+                output.clear();
+        }
+        else if (path == "/..")
+        {
+            path = "/";
+            std::size_t slash = output.find_last_of('/');
+            if (slash != std::string::npos)
+                output.erase(slash);
+            else
+                output.clear();
+        }
+        else if (path == "." || path == "..")
+        {
+            path.clear();
+        }
+        else
+        {
+            std::size_t next = path.find('/', 1);
+            if (next == std::string::npos)
+            {
+                output += path;
+                path.clear();
+            }
+            else
+            {
+                output += path.substr(0, next);
+                path.erase(0, next);
+            }
+        }
+    }
+    return output;
+}
+
+std::string merge_paths(const std::string &base_path, const std::string &relative_path, bool base_has_authority)
+{
+    if (base_has_authority && base_path.empty())
+    {
+        return "/" + relative_path;
+    }
+
+    const std::size_t slash = base_path.find_last_of('/');
+    if (slash == std::string::npos)
+    {
+        return relative_path;
+    }
+
+    return base_path.substr(0, slash + 1) + relative_path;
+}
+
+struct parsed_uri
+{
+    bool absolute = false;
+    bool has_authority = false;
+    bool has_authentication = false;
+    bool has_port = false;
+    bool has_query = false;
+    bool has_fragment = false;
+
+    std::string scheme;
+    std::string username;
+    std::string password;
+    std::string host;
+    std::size_t port = 0;
+    std::string query;
+    std::string fragment;
+    std::string path;
+};
+
+parsed_uri parse_uri_string(const std::string &uri_string)
+{
+    parsed_uri parsed;
+    std::string work = uri_string;
+
+    const std::size_t fragment_pos = work.find('#');
+    if (fragment_pos != std::string::npos)
+    {
+        parsed.fragment = work.substr(fragment_pos + 1);
+        parsed.has_fragment = true;
+        work = work.substr(0, fragment_pos);
+    }
+
+    const std::size_t query_pos = work.find('?');
+    if (query_pos != std::string::npos)
+    {
+        parsed.query = work.substr(query_pos + 1);
+        parsed.has_query = true;
+        work = work.substr(0, query_pos);
+    }
+
+    const std::size_t first_special = work.find_first_of("/?#");
+    const std::string maybe_scheme = (first_special == std::string::npos) ? work : work.substr(0, first_special);
+
+    std::size_t scheme_pos = std::string::npos;
+    if (looks_like_scheme(maybe_scheme))
+    {
+        scheme_pos = maybe_scheme.find(':');
+    }
+
+    std::string rest = work;
+    if (scheme_pos != std::string::npos)
+    {
+        parsed.scheme = work.substr(0, scheme_pos);
+        parsed.absolute = true;
+        rest = work.substr(scheme_pos + 1);
+    }
+
+    if (starts_with(rest, "//"))
+    {
+        parsed.has_authority = true;
+        const std::size_t authority_begin = 2;
+        const std::size_t authority_end = rest.find_first_of("/?", authority_begin);
+
+        const std::string authority = rest.substr(
+            authority_begin,
+            authority_end == std::string::npos ? std::string::npos : authority_end - authority_begin);
+
+        rest = (authority_end == std::string::npos) ? std::string() : rest.substr(authority_end);
+
+        const parsed_authority auth = parse_authority(authority);
+        parsed.has_authentication = auth.authentication;
+        parsed.has_port = auth.port_present;
+        parsed.username = auth.username;
+        parsed.password = auth.password;
+        parsed.host = auth.host;
+        parsed.port = auth.port;
+    }
+
+    parsed.path = rest;
+    return parsed;
+}
+
+std::string relative_path_from_base(const std::string &base_path, const std::string &target_path, bool base_has_authority)
+{
+    if (target_path == base_path)
+    {
+        return std::string();
+    }
+
+    std::string base_dir = base_path;
+    if (base_dir.empty())
+    {
+        if (base_has_authority)
+        {
+            base_dir = "/";
+        }
+    }
+    else if (base_dir.back() != '/')
+    {
+        const std::size_t slash = base_dir.find_last_of('/');
+        if (slash != std::string::npos)
+        {
+            base_dir.erase(slash + 1);
+        }
+        else
+        {
+            base_dir.clear();
+        }
+    }
+
+    auto split = [](const std::string &p)
+    {
+        std::vector<std::string> out;
+        std::size_t start = 0;
+        while (start <= p.size())
+        {
+            std::size_t pos = p.find('/', start);
+            if (pos == std::string::npos)
+            {
+                out.push_back(p.substr(start));
+                break;
+            }
+            out.push_back(p.substr(start, pos - start));
+            start = pos + 1;
+            if (start == p.size())
+            {
+                out.push_back("");
+                break;
+            }
+        }
+        return out;
+    };
+
+    const std::vector<std::string> base_segments = split(base_dir);
+    const std::vector<std::string> target_segments = split(target_path);
+
+    std::size_t common = 0;
+    while (common < base_segments.size() &&
+           common < target_segments.size() &&
+           base_segments[common] == target_segments[common])
+    {
+        ++common;
+    }
+
+    std::vector<std::string> result_segments;
+
+    for (std::size_t i = common; i < base_segments.size(); ++i)
+    {
+        if (!base_segments[i].empty())
+        {
+            result_segments.push_back("..");
+        }
+    }
+
+    for (std::size_t i = common; i < target_segments.size(); ++i)
+    {
+        result_segments.push_back(target_segments[i]);
+    }
+
+    std::string result;
+    for (std::size_t i = 0; i < result_segments.size(); ++i)
+    {
+        if (i != 0)
+        {
+            result.push_back('/');
+        }
+        result += result_segments[i];
+    }
+
+    return result;
+}
+
+std::string normalize_for_compare(const std::string &path)
+{
+    return remove_dot_segments(path);
+}
+} // namespace
+
+uri::uri() = default;
 
 uri::uri(const std::string &uri_string)
-    : path_(uri_string)
 {
+    const parsed_uri parsed = parse_uri_string(uri_string);
+
+    absolute_ = parsed.absolute;
+    has_authority_ = parsed.has_authority;
+    has_authentication_ = parsed.has_authentication;
+    has_port_ = parsed.has_port;
+    has_query_ = parsed.has_query;
+    has_fragment_ = parsed.has_fragment;
+
+    scheme_ = parsed.scheme;
+    username_ = parsed.username;
+    password_ = parsed.password;
+    host_ = parsed.host;
+    port_ = parsed.port;
+    query_ = parsed.query;
+    fragment_ = parsed.fragment;
+    path_ = xlnt::path(parsed.path);
 }
 
-std::string uri::to_string() const
+uri::uri(const uri &base, const uri &relative)
 {
-    return path_.string();
+    if (relative.absolute_)
+    {
+        *this = relative;
+        return;
+    }
+
+    absolute_ = base.absolute_;
+    scheme_ = base.scheme_;
+
+    if (relative.has_authority_)
+    {
+        has_authority_ = true;
+        has_authentication_ = relative.has_authentication_;
+        has_port_ = relative.has_port_;
+        username_ = relative.username_;
+        password_ = relative.password_;
+        host_ = relative.host_;
+        port_ = relative.port_;
+        path_ = xlnt::path(remove_dot_segments(relative.path_.string()));
+        has_query_ = relative.has_query_;
+        query_ = relative.query_;
+    }
+    else if (relative.path_.string().empty())
+    {
+        has_authority_ = base.has_authority_;
+        has_authentication_ = base.has_authentication_;
+        has_port_ = base.has_port_;
+        username_ = base.username_;
+        password_ = base.password_;
+        host_ = base.host_;
+        port_ = base.port_;
+        path_ = base.path_;
+
+        if (relative.has_query_)
+        {
+            has_query_ = true;
+            query_ = relative.query_;
+        }
+        else
+        {
+            has_query_ = base.has_query_;
+            query_ = base.query_;
+        }
+    }
+    else
+    {
+        has_authority_ = base.has_authority_;
+        has_authentication_ = base.has_authentication_;
+        has_port_ = base.has_port_;
+        username_ = base.username_;
+        password_ = base.password_;
+        host_ = base.host_;
+        port_ = base.port_;
+
+        const std::string rel_path = relative.path_.string();
+
+        if (!rel_path.empty() && rel_path.front() == '/')
+        {
+            path_ = xlnt::path(remove_dot_segments(rel_path));
+        }
+        else
+        {
+            const std::string base_path = base.path_.string();
+            const std::string merged = merge_paths(base_path, rel_path, base.has_authority_);
+            path_ = xlnt::path(remove_dot_segments(merged));
+        }
+
+        has_query_ = relative.has_query_;
+        query_ = relative.query_;
+    }
+
+    has_fragment_ = relative.has_fragment_;
+    fragment_ = relative.fragment_;
 }
 
-const path &uri::path() const
+uri::uri(const uri &base, const xlnt::path &relative)
+{
+    uri ref;
+    ref.absolute_ = false;
+    ref.has_authority_ = false;
+    ref.path_ = relative;
+
+    *this = uri(base, ref);
+}
+
+bool uri::is_relative() const
+{
+    return !absolute_;
+}
+
+bool uri::is_absolute() const
+{
+    return absolute_;
+}
+
+std::string uri::scheme() const
+{
+    return scheme_;
+}
+
+std::string uri::authority() const
+{
+    if (!has_authority_)
+    {
+        return std::string();
+    }
+
+    std::string result;
+
+    if (has_authentication_)
+    {
+        result += username_;
+        if (!password_.empty())
+        {
+            result += ':';
+            result += password_;
+        }
+        result += '@';
+    }
+
+    result += bracket_ipv6_if_needed(host_);
+
+    if (has_port_)
+    {
+        result += ':';
+        result += std::to_string(port_);
+    }
+
+    return result;
+}
+
+bool uri::has_authentication() const
+{
+    return has_authentication_;
+}
+
+std::string uri::authentication() const
+{
+    if (!has_authentication_)
+    {
+        return std::string();
+    }
+
+    if (password_.empty())
+    {
+        return username_;
+    }
+
+    return username_ + ":" + password_;
+}
+
+std::string uri::username() const
+{
+    return username_;
+}
+
+std::string uri::password() const
+{
+    return password_;
+}
+
+std::string uri::host() const
+{
+    return host_;
+}
+
+bool uri::has_port() const
+{
+    return has_port_;
+}
+
+std::size_t uri::port() const
+{
+    return port_;
+}
+
+const class path &uri::path() const
 {
     return path_;
 }
 
+bool uri::has_query() const
+{
+    return has_query_;
+}
+
+std::string uri::query() const
+{
+    return query_;
+}
+
+bool uri::has_fragment() const
+{
+    return has_fragment_;
+}
+
+std::string uri::fragment() const
+{
+    return fragment_;
+}
+
+std::string uri::to_string() const
+{
+    std::string result;
+
+    if (!scheme_.empty())
+    {
+        result += scheme_;
+        result += ':';
+    }
+
+    if (has_authority_)
+    {
+        result += "//";
+        result += authority();
+    }
+
+    result += path_.string();
+
+    if (has_query_)
+    {
+        result += '?';
+        result += query_;
+    }
+
+    if (has_fragment_)
+    {
+        result += '#';
+        result += fragment_;
+    }
+
+    return result;
+}
+
+uri uri::make_absolute(const uri &base)
+{
+    if (is_absolute())
+    {
+        return *this;
+    }
+
+    return uri(base, *this);
+}
+
+uri uri::make_reference(const uri &base)
+{
+    if (!is_absolute())
+    {
+        return *this;
+    }
+
+    if (lower_copy(scheme_) != lower_copy(base.scheme_) ||
+        has_authority_ != base.has_authority_ ||
+        has_authentication_ != base.has_authentication_ ||
+        has_port_ != base.has_port_ ||
+        lower_copy(host_) != lower_copy(base.host_) ||
+        username_ != base.username_ ||
+        password_ != base.password_ ||
+        port_ != base.port_)
+    {
+        return *this;
+    }
+
+    uri result;
+    result.absolute_ = false;
+    result.has_authority_ = false;
+    result.has_authentication_ = false;
+    result.has_port_ = false;
+    result.has_query_ = has_query_;
+    result.query_ = query_;
+    result.has_fragment_ = has_fragment_;
+    result.fragment_ = fragment_;
+
+    result.path_ = xlnt::path(relative_path_from_base(base.path_.string(), path_.string(), base.has_authority_));
+    return result;
+}
+
+bool uri::has_authority() const
+{
+    return has_authority_;
+}
+
 bool uri::operator==(const uri &other) const
 {
-    return to_string() == other.to_string();
+    return absolute_ == other.absolute_ &&
+           has_authority_ == other.has_authority_ &&
+           has_authentication_ == other.has_authentication_ &&
+           has_port_ == other.has_port_ &&
+           has_query_ == other.has_query_ &&
+           has_fragment_ == other.has_fragment_ &&
+           lower_copy(scheme_) == lower_copy(other.scheme_) &&
+           username_ == other.username_ &&
+           password_ == other.password_ &&
+           lower_copy(host_) == lower_copy(other.host_) &&
+           port_ == other.port_ &&
+           query_ == other.query_ &&
+           fragment_ == other.fragment_ &&
+           normalize_for_compare(path_.string()) == normalize_for_compare(other.path_.string());
 }
 
 bool uri::operator!=(const uri &other) const

--- a/source/packaging/uri.cpp
+++ b/source/packaging/uri.cpp
@@ -98,17 +98,6 @@ bool parse_port(const std::string &s, std::size_t &port)
     return true;
 }
 
-std::string bracket_ipv6_if_needed(const std::string &host)
-{
-    if (host.find(':') != std::string::npos &&
-        !(host.size() >= 2 && host.front() == '[' && host.back() == ']'))
-    {
-        return "[" + host + "]";
-    }
-
-    return host;
-}
-
 struct parsed_authority
 {
     bool present = false;

--- a/source/packaging/uri.cpp
+++ b/source/packaging/uri.cpp
@@ -118,6 +118,7 @@ struct parsed_authority
     std::string password;
     std::string host;
     std::size_t port = 0;
+    bool ipv6 = false;
 };
 
 parsed_authority parse_authority(const std::string &authority)
@@ -153,6 +154,7 @@ parsed_authority parse_authority(const std::string &authority)
 
     if (!hostport.empty() && hostport.front() == '[')
     {
+        result.ipv6 = true;
         const std::size_t close = hostport.find(']');
         if (close != std::string::npos)
         {
@@ -283,6 +285,7 @@ struct parsed_uri
     bool has_port = false;
     bool has_query = false;
     bool has_fragment = false;
+    bool ipv6 = false;
 
     std::string scheme;
     std::string username;
@@ -345,6 +348,7 @@ parsed_uri parse_uri_string(const std::string &uri_string)
         rest = (authority_end == std::string::npos) ? std::string() : rest.substr(authority_end);
 
         const parsed_authority auth = parse_authority(authority);
+        parsed.ipv6 = auth.ipv6;
         parsed.has_authentication = auth.authentication;
         parsed.has_port = auth.port_present;
         parsed.username = auth.username;
@@ -461,6 +465,7 @@ uri::uri(const std::string &uri_string)
 
     absolute_ = parsed.absolute;
     has_authority_ = parsed.has_authority;
+    ipv6_ = parsed.ipv6;
     has_authentication_ = parsed.has_authentication;
     has_port_ = parsed.has_port;
     has_query_ = parsed.has_query;
@@ -549,6 +554,7 @@ uri::uri(const uri &base, const uri &relative)
         query_ = relative.query_;
     }
 
+    ipv6_ = has_authority_ ? (relative.has_authority_ ? relative.ipv6_ : base.ipv6_) : false;
     has_fragment_ = relative.has_fragment_;
     fragment_ = relative.fragment_;
 }
@@ -559,6 +565,7 @@ uri::uri(const uri &base, const xlnt::path &relative)
     ref.absolute_ = false;
     ref.has_authority_ = false;
     ref.path_ = relative;
+    ref.ipv6_ = base.ipv6_;
 
     *this = uri(base, ref);
 }
@@ -598,7 +605,14 @@ std::string uri::authority() const
         result += '@';
     }
 
-    result += bracket_ipv6_if_needed(host_);
+    if (ipv6_)
+    {
+        result += '[' + host_ + ']';
+    }
+    else
+    {
+        result += host_;
+    }
 
     if (has_port_)
     {

--- a/tests/helpers/internal/xml_helper.hpp
+++ b/tests/helpers/internal/xml_helper.hpp
@@ -348,7 +348,7 @@ public:
                         << static_cast<int>(rel.type())
                         << ':' << static_cast<int>(rel.target_mode())
                         << ':' << rel.source().path().string()
-                        << ':' << rel.target().path().string() << '\n';
+                        << ':' << rel.target().to_string() << '\n';
                 }
             }
             std::cout << "\nRight:\n";
@@ -362,7 +362,7 @@ public:
                         << ':' << static_cast<int>(rel.type())
                         << ':' << static_cast<int>(rel.target_mode())
                         << ':' << rel.source().path().string()
-                        << ':' << rel.target().path().string() << '\n';
+                        << ':' << rel.target().to_string() << '\n';
                 }
             }
             return false;

--- a/tests/utils/uri_test_suite.cpp
+++ b/tests/utils/uri_test_suite.cpp
@@ -1,0 +1,374 @@
+// Copyright (c) 2014-2022 Thomas Fussell
+// Copyright (c) 2024-2026 xlnt-community
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE
+//
+// @license: http://www.opensource.org/licenses/mit-license.php
+// @author: see AUTHORS file
+
+#include <helpers/test_suite.hpp>
+
+#include <xlnt/packaging/uri.hpp>
+
+class uri_test_suite : public test_suite
+{
+public:
+    uri_test_suite()
+    {
+        register_test(test_default_ctor);
+        register_test(test_parse_relative_path);
+        register_test(test_parse_absolute_basic);
+        register_test(test_parse_with_auth_query_fragment);
+        register_test(test_parse_ipv6_host);
+        register_test(test_parse_empty_query_and_fragment);
+        register_test(test_parse_username_only);
+        register_test(test_parse_network_path_reference);
+        register_test(test_parse_empty_authority_file_uri);
+        register_test(test_parse_query_only);
+        register_test(test_parse_fragment_only);
+        register_test(test_percent_encoding_preserved);
+        register_test(test_to_string_round_trip);
+        register_test(test_equality);
+        register_test(test_make_absolute_simple);
+        register_test(test_make_absolute_parent_segments);
+        register_test(test_make_absolute_root_relative);
+        register_test(test_make_reference_simple);
+        register_test(test_make_reference_different_authority);
+        register_test(test_make_reference_same_uri);
+        register_test(test_rfc3986_resolution_examples);
+    }
+
+    void test_default_ctor()
+    {
+        xlnt::uri u;
+
+        xlnt_assert(u.is_relative());
+        xlnt_assert(!u.is_absolute());
+        xlnt_assert_equals(u.scheme(), "");
+        xlnt_assert_equals(u.authority(), "");
+        xlnt_assert(!u.has_authentication());
+        xlnt_assert_equals(u.authentication(), "");
+        xlnt_assert_equals(u.username(), "");
+        xlnt_assert_equals(u.password(), "");
+        xlnt_assert_equals(u.host(), "");
+        xlnt_assert(!u.has_port());
+        xlnt_assert_equals(u.port(), std::size_t(0));
+        xlnt_assert_equals(u.path().string(), "");
+        xlnt_assert(!u.has_query());
+        xlnt_assert_equals(u.query(), "");
+        xlnt_assert(!u.has_fragment());
+        xlnt_assert_equals(u.fragment(), "");
+        xlnt_assert_equals(u.to_string(), "");
+    }
+
+    void test_parse_relative_path()
+    {
+        xlnt::uri u("folder/file.txt");
+
+        xlnt_assert(u.is_relative());
+        xlnt_assert_equals(u.scheme(), "");
+        xlnt_assert_equals(u.authority(), "");
+        xlnt_assert_equals(u.path().string(), "folder/file.txt");
+        xlnt_assert_equals(u.to_string(), "folder/file.txt");
+        xlnt_assert(!u.has_query());
+        xlnt_assert(!u.has_fragment());
+    }
+
+    void test_parse_absolute_basic()
+    {
+        xlnt::uri u("http://example.com/document");
+
+        xlnt_assert(u.is_absolute());
+        xlnt_assert_equals(u.scheme(), "http");
+        xlnt_assert_equals(u.authority(), "example.com");
+        xlnt_assert_equals(u.host(), "example.com");
+        xlnt_assert(!u.has_authentication());
+        xlnt_assert(!u.has_port());
+        xlnt_assert_equals(u.path().string(), "/document");
+        xlnt_assert_equals(u.to_string(), "http://example.com/document");
+    }
+
+    void test_parse_with_auth_query_fragment()
+    {
+        xlnt::uri u("http://user:pass@example.com:80/document?v=1&x=3#abc");
+
+        xlnt_assert(u.is_absolute());
+        xlnt_assert_equals(u.scheme(), "http");
+        xlnt_assert(u.has_authentication());
+        xlnt_assert_equals(u.authentication(), "user:pass");
+        xlnt_assert_equals(u.username(), "user");
+        xlnt_assert_equals(u.password(), "pass");
+        xlnt_assert_equals(u.host(), "example.com");
+        xlnt_assert(u.has_port());
+        xlnt_assert_equals(u.port(), std::size_t(80));
+        xlnt_assert_equals(u.authority(), "user:pass@example.com:80");
+        xlnt_assert_equals(u.path().string(), "/document");
+        xlnt_assert(u.has_query());
+        xlnt_assert_equals(u.query(), "v=1&x=3");
+        xlnt_assert(u.has_fragment());
+        xlnt_assert_equals(u.fragment(), "abc");
+        xlnt_assert_equals(u.to_string(), "http://user:pass@example.com:80/document?v=1&x=3#abc");
+    }
+
+    void test_parse_ipv6_host()
+    {
+        xlnt::uri u("https://[::1]:443/index.html");
+
+        xlnt_assert(u.is_absolute());
+        xlnt_assert_equals(u.scheme(), "https");
+        xlnt_assert_equals(u.host(), "::1");
+        xlnt_assert(u.has_port());
+        xlnt_assert_equals(u.port(), std::size_t(443));
+        xlnt_assert_equals(u.authority(), "[::1]:443");
+        xlnt_assert_equals(u.to_string(), "https://[::1]:443/index.html");
+    }
+
+    void test_parse_empty_query_and_fragment()
+    {
+        xlnt::uri u1("http://example.com/document?");
+        xlnt_assert(u1.has_query());
+        xlnt_assert_equals(u1.query(), "");
+        xlnt_assert_equals(u1.to_string(), "http://example.com/document?");
+
+        xlnt::uri u2("http://example.com/document#");
+        xlnt_assert(u2.has_fragment());
+        xlnt_assert_equals(u2.fragment(), "");
+        xlnt_assert_equals(u2.to_string(), "http://example.com/document#");
+
+        xlnt::uri u3("http://example.com/document?#");
+        xlnt_assert(u3.has_query());
+        xlnt_assert(u3.has_fragment());
+        xlnt_assert_equals(u3.query(), "");
+        xlnt_assert_equals(u3.fragment(), "");
+        xlnt_assert_equals(u3.to_string(), "http://example.com/document?#");
+    }
+
+    void test_parse_username_only()
+    {
+        xlnt::uri u("http://user@example.com");
+
+        xlnt_assert(u.is_absolute());
+        xlnt_assert(u.has_authentication());
+        xlnt_assert_equals(u.username(), "user");
+        xlnt_assert_equals(u.password(), "");
+        xlnt_assert_equals(u.authentication(), "user");
+        xlnt_assert_equals(u.authority(), "user@example.com");
+        xlnt_assert_equals(u.to_string(), "http://user@example.com");
+    }
+
+    void test_parse_network_path_reference()
+    {
+        xlnt::uri u("//example.com/path");
+
+        xlnt_assert(u.is_relative());
+        xlnt_assert_equals(u.scheme(), "");
+        xlnt_assert_equals(u.authority(), "example.com");
+        xlnt_assert_equals(u.host(), "example.com");
+        xlnt_assert_equals(u.path().string(), "/path");
+        xlnt_assert_equals(u.to_string(), "//example.com/path");
+    }
+
+    void test_parse_empty_authority_file_uri()
+    {
+        xlnt::uri u("file:///tmp/test.txt");
+
+        xlnt_assert(u.is_absolute());
+        xlnt_assert_equals(u.scheme(), "file");
+        xlnt_assert_equals(u.authority(), "");
+        xlnt_assert_equals(u.host(), "");
+        xlnt_assert_equals(u.path().string(), "/tmp/test.txt");
+        xlnt_assert_equals(u.to_string(), "file:///tmp/test.txt");
+    }
+
+    void test_parse_query_only()
+    {
+        xlnt::uri u("?x=1");
+
+        xlnt_assert(u.is_relative());
+        xlnt_assert_equals(u.path().string(), "");
+        xlnt_assert(u.has_query());
+        xlnt_assert_equals(u.query(), "x=1");
+        xlnt_assert(!u.has_fragment());
+        xlnt_assert_equals(u.to_string(), "?x=1");
+    }
+
+    void test_parse_fragment_only()
+    {
+        xlnt::uri u("#frag");
+
+        xlnt_assert(u.is_relative());
+        xlnt_assert_equals(u.path().string(), "");
+        xlnt_assert(!u.has_query());
+        xlnt_assert(u.has_fragment());
+        xlnt_assert_equals(u.fragment(), "frag");
+        xlnt_assert_equals(u.to_string(), "#frag");
+    }
+
+    void test_percent_encoding_preserved()
+    {
+        xlnt::uri u("http://example.com/a%20b");
+
+        xlnt_assert_equals(u.path().string(), "/a%20b");
+        xlnt_assert_equals(u.to_string(), "http://example.com/a%20b");
+    }
+
+    void test_to_string_round_trip()
+    {
+        const std::string input = "https://user:pw@example.org:8443/a/b/c?x=1&y=2#frag";
+        xlnt::uri u(input);
+
+        xlnt_assert_equals(u.to_string(), input);
+
+        xlnt::uri round_trip(u.to_string());
+        xlnt_assert_equals(round_trip.to_string(), input);
+        xlnt_assert(round_trip == u);
+    }
+
+    void test_equality()
+    {
+        xlnt::uri a("http://example.com/a");
+        xlnt::uri b("http://example.com/a");
+        xlnt::uri c("http://example.com/b");
+
+        xlnt_assert(a == b);
+        xlnt_assert(!(a != b));
+        xlnt_assert(a != c);
+        xlnt_assert(!(a == c));
+    }
+
+    void test_make_absolute_simple()
+    {
+        xlnt::uri base("http://example.com/dir/index.html");
+        xlnt::uri relative("child.txt");
+
+        xlnt::uri abs = relative.make_absolute(base);
+
+        xlnt_assert(abs.is_absolute());
+        xlnt_assert_equals(abs.scheme(), "http");
+        xlnt_assert_equals(abs.authority(), "example.com");
+        xlnt_assert_equals(abs.path().string(), "/dir/child.txt");
+        xlnt_assert_equals(abs.to_string(), "http://example.com/dir/child.txt");
+    }
+
+    void test_make_absolute_parent_segments()
+    {
+        xlnt::uri base("http://example.com/a/b/c/");
+        xlnt::uri relative("../d/e.txt");
+
+        xlnt::uri abs = relative.make_absolute(base);
+
+        xlnt_assert(abs.is_absolute());
+        xlnt_assert_equals(abs.path().string(), "/a/b/d/e.txt");
+        xlnt_assert_equals(abs.to_string(), "http://example.com/a/b/d/e.txt");
+    }
+
+    void test_make_absolute_root_relative()
+    {
+        xlnt::uri base("http://example.com/a/b/c");
+        xlnt::uri relative("/root/file.txt");
+
+        xlnt::uri abs = relative.make_absolute(base);
+
+        xlnt_assert(abs.is_absolute());
+        xlnt_assert_equals(abs.path().string(), "/root/file.txt");
+        xlnt_assert_equals(abs.to_string(), "http://example.com/root/file.txt");
+    }
+
+    void test_make_reference_simple()
+    {
+        xlnt::uri base("http://example.com/a/b/");
+        xlnt::uri abs("http://example.com/a/b/c/d.txt");
+
+        xlnt::uri ref = abs.make_reference(base);
+
+        xlnt_assert(ref.is_relative());
+        xlnt_assert_equals(ref.scheme(), "");
+        xlnt_assert_equals(ref.authority(), "");
+        xlnt_assert_equals(ref.path().string(), "c/d.txt");
+        xlnt_assert_equals(ref.to_string(), "c/d.txt");
+    }
+
+    void test_make_reference_different_authority()
+    {
+        xlnt::uri base("http://example.com/a/b/");
+        xlnt::uri abs("http://other.example.com/a/b/c.txt");
+
+        xlnt::uri ref = abs.make_reference(base);
+
+        xlnt_assert(ref.is_absolute());
+        xlnt_assert_equals(ref.to_string(), "http://other.example.com/a/b/c.txt");
+    }
+
+    void test_make_reference_same_uri()
+    {
+        xlnt::uri base("http://example.com/a/b/");
+        xlnt::uri abs("http://example.com/a/b/");
+
+        xlnt::uri ref = abs.make_reference(base);
+
+        xlnt_assert(ref.is_relative());
+        xlnt_assert_equals(ref.path().string(), "");
+        xlnt_assert_equals(ref.to_string(), "");
+    }
+
+    void test_rfc3986_resolution_examples()
+    {
+        xlnt::uri base("http://a/b/c/d;p?q");
+
+        struct test_case
+        {
+            const char *relative;
+            const char *expected;
+        };
+
+        const test_case cases[] =
+        {
+            {"g:h", "g:h"},
+            {"g", "http://a/b/c/g"},
+            {"./g", "http://a/b/c/g"},
+            {"g/", "http://a/b/c/g/"},
+            {"/g", "http://a/g"},
+            {"//g", "http://g"},
+            {"?y", "http://a/b/c/d;p?y"},
+            {"g?y", "http://a/b/c/g?y"},
+            {"#s", "http://a/b/c/d;p?q#s"},
+            {"g#s", "http://a/b/c/g#s"},
+            {"g?y#s", "http://a/b/c/g?y#s"},
+            {";x", "http://a/b/c/;x"},
+            {"g;x", "http://a/b/c/g;x"},
+            {"", "http://a/b/c/d;p?q"},
+            {".", "http://a/b/c/"},
+            {"./", "http://a/b/c/"},
+            {"..", "http://a/b/"},
+            {"../", "http://a/b/"},
+            {"../g", "http://a/b/g"},
+            {"../..", "http://a/"},
+            {"../../g", "http://a/g"}
+        };
+
+        for (const auto &test : cases)
+        {
+            xlnt::uri relative(test.relative);
+            xlnt::uri resolved = relative.make_absolute(base);
+            xlnt_assert_equals(resolved.to_string(), std::string(test.expected));
+        }
+    }
+};
+
+static uri_test_suite x;

--- a/tests/utils/uri_test_suite.cpp
+++ b/tests/utils/uri_test_suite.cpp
@@ -57,6 +57,7 @@ public:
         register_test(test_make_absolute_relative_passthrough);
         register_test(test_parse_invalid_port_empty);
         register_test(test_parse_invalid_port_nondigit);
+        register_test(test_uri_edge_coverage);
     }
 
     void test_default_ctor()
@@ -439,6 +440,31 @@ public:
         xlnt_assert(!u.has_port());
         xlnt_assert_equals(u.host(), "example.com:abc");
         xlnt_assert_equals(u.to_string(), "http://example.com:abc/path");
+    }
+
+    void test_uri_edge_coverage()
+    {
+        xlnt_assert(!xlnt::uri("http://example.com:999999999999999999999999/path").has_port());
+
+        xlnt::uri bad_ipv6("http://[::1/path");
+        xlnt_assert(bad_ipv6.has_authority());
+
+        xlnt::uri base_empty("");
+
+        xlnt_assert_equals(xlnt::uri("/./g").make_absolute(base_empty).to_string(), "/g");
+        xlnt_assert_equals(xlnt::uri("/.").make_absolute(base_empty).to_string(), "/");
+        xlnt_assert_equals(xlnt::uri("/../g").make_absolute(base_empty).to_string(), "/g");
+        xlnt_assert_equals(xlnt::uri("/..").make_absolute(base_empty).to_string(), "/");
+
+
+        xlnt_assert_equals(xlnt::uri("child").make_absolute(xlnt::uri("http://example.com")).to_string(), "http://example.com/child");
+        xlnt_assert_equals(xlnt::uri("child").make_absolute(xlnt::uri("urn:base")).to_string(), "urn:child");
+
+        xlnt::uri ref1 = xlnt::uri("http://example.com/a").make_reference(xlnt::uri("http://example.com"));
+        xlnt_assert_equals(ref1.to_string(), "a");
+
+        xlnt::uri ref2 = xlnt::uri("urn:target").make_reference(xlnt::uri("urn:base"));
+        xlnt_assert_equals(ref2.to_string(), "target");
     }
 };
 

--- a/tests/utils/uri_test_suite.cpp
+++ b/tests/utils/uri_test_suite.cpp
@@ -32,6 +32,7 @@ public:
     uri_test_suite()
     {
         register_test(test_default_ctor);
+        register_test(test_make_absolute_path_overload);
         register_test(test_parse_relative_path);
         register_test(test_parse_absolute_basic);
         register_test(test_parse_with_auth_query_fragment);
@@ -52,6 +53,10 @@ public:
         register_test(test_make_reference_different_authority);
         register_test(test_make_reference_same_uri);
         register_test(test_rfc3986_resolution_examples);
+        register_test(test_make_reference_relative_passthrough);
+        register_test(test_make_absolute_relative_passthrough);
+        register_test(test_parse_invalid_port_empty);
+        register_test(test_parse_invalid_port_nondigit);
     }
 
     void test_default_ctor()
@@ -75,6 +80,16 @@ public:
         xlnt_assert(!u.has_fragment());
         xlnt_assert_equals(u.fragment(), "");
         xlnt_assert_equals(u.to_string(), "");
+        xlnt_assert(!u.has_authority());
+    }
+
+    void test_make_absolute_path_overload()
+    {
+        xlnt::uri base("http://example.com/dir/index.html");
+        xlnt::uri u(base, xlnt::path("child.txt"));
+
+        xlnt_assert(u.is_absolute());
+        xlnt_assert_equals(u.to_string(), "http://example.com/dir/child.txt");
     }
 
     void test_parse_relative_path()
@@ -88,6 +103,7 @@ public:
         xlnt_assert_equals(u.to_string(), "folder/file.txt");
         xlnt_assert(!u.has_query());
         xlnt_assert(!u.has_fragment());
+        xlnt_assert(!u.has_authority());
     }
 
     void test_parse_absolute_basic()
@@ -102,6 +118,7 @@ public:
         xlnt_assert(!u.has_port());
         xlnt_assert_equals(u.path().string(), "/document");
         xlnt_assert_equals(u.to_string(), "http://example.com/document");
+        xlnt_assert(u.has_authority());
     }
 
     void test_parse_with_auth_query_fragment()
@@ -124,6 +141,7 @@ public:
         xlnt_assert(u.has_fragment());
         xlnt_assert_equals(u.fragment(), "abc");
         xlnt_assert_equals(u.to_string(), "http://user:pass@example.com:80/document?v=1&x=3#abc");
+        xlnt_assert(u.has_authority());
     }
 
     void test_parse_ipv6_host()
@@ -141,12 +159,18 @@ public:
 
     void test_parse_empty_query_and_fragment()
     {
+        xlnt::uri plain("http://example.com/document");
+        xlnt_assert(!plain.has_query());
+        xlnt_assert(!plain.has_fragment());
+
         xlnt::uri u1("http://example.com/document?");
         xlnt_assert(u1.has_query());
+        xlnt_assert(!u1.has_fragment());
         xlnt_assert_equals(u1.query(), "");
         xlnt_assert_equals(u1.to_string(), "http://example.com/document?");
 
         xlnt::uri u2("http://example.com/document#");
+        xlnt_assert(!u2.has_query());
         xlnt_assert(u2.has_fragment());
         xlnt_assert_equals(u2.fragment(), "");
         xlnt_assert_equals(u2.to_string(), "http://example.com/document#");
@@ -182,6 +206,7 @@ public:
         xlnt_assert_equals(u.host(), "example.com");
         xlnt_assert_equals(u.path().string(), "/path");
         xlnt_assert_equals(u.to_string(), "//example.com/path");
+        xlnt_assert(u.has_authority());
     }
 
     void test_parse_empty_authority_file_uri()
@@ -194,6 +219,7 @@ public:
         xlnt_assert_equals(u.host(), "");
         xlnt_assert_equals(u.path().string(), "/tmp/test.txt");
         xlnt_assert_equals(u.to_string(), "file:///tmp/test.txt");
+        xlnt_assert(u.has_authority());
     }
 
     void test_parse_query_only()
@@ -368,6 +394,51 @@ public:
             xlnt::uri resolved = relative.make_absolute(base);
             xlnt_assert_equals(resolved.to_string(), std::string(test.expected));
         }
+    }
+
+    void test_make_reference_relative_passthrough()
+    {
+        xlnt::uri base("http://example.com/a/b/");
+        xlnt::uri rel("c/d.txt");
+
+        xlnt::uri ref = rel.make_reference(base);
+
+        xlnt_assert(ref.is_relative());
+        xlnt_assert_equals(ref.to_string(), "c/d.txt");
+    }
+
+    void test_make_absolute_relative_passthrough()
+    {
+        xlnt::uri base("http://example.com/a/b/");
+        xlnt::uri absolute_relative("https://other.example.com/x?y=1#f");
+
+        xlnt::uri resolved(base, absolute_relative);
+
+        xlnt_assert(resolved.is_absolute());
+        xlnt_assert_equals(resolved.to_string(), absolute_relative.to_string());
+        xlnt_assert(resolved == absolute_relative);
+    }
+
+    void test_parse_invalid_port_empty()
+    {
+        xlnt::uri u("http://example.com:/path");
+
+        xlnt_assert(u.is_absolute());
+        xlnt_assert(u.has_authority());
+        xlnt_assert(!u.has_port());
+        xlnt_assert_equals(u.host(), "example.com:");
+        xlnt_assert_equals(u.to_string(), "http://example.com:/path");
+    }
+
+    void test_parse_invalid_port_nondigit()
+    {
+        xlnt::uri u("http://example.com:abc/path");
+
+        xlnt_assert(u.is_absolute());
+        xlnt_assert(u.has_authority());
+        xlnt_assert(!u.has_port());
+        xlnt_assert_equals(u.host(), "example.com:abc");
+        xlnt_assert_equals(u.to_string(), "http://example.com:abc/path");
     }
 };
 

--- a/tests/utils/uri_test_suite.cpp
+++ b/tests/utils/uri_test_suite.cpp
@@ -386,7 +386,7 @@ public:
             {"../", "http://a/b/"},
             {"../g", "http://a/b/g"},
             {"../..", "http://a/"},
-            {"../../g", "http://a/g"}
+            {"../../g", "http://a/g"},
         };
 
         for (const auto &test : cases)
@@ -452,10 +452,14 @@ public:
         xlnt::uri base_empty("");
 
         xlnt_assert_equals(xlnt::uri("/./g").make_absolute(base_empty).to_string(), "/g");
+        xlnt_assert_equals(xlnt::uri("./g").make_absolute(base_empty).to_string(), "g");
         xlnt_assert_equals(xlnt::uri("/.").make_absolute(base_empty).to_string(), "/");
+        xlnt_assert_equals(xlnt::uri("./").make_absolute(base_empty).to_string(), "");
         xlnt_assert_equals(xlnt::uri("/../g").make_absolute(base_empty).to_string(), "/g");
         xlnt_assert_equals(xlnt::uri("/..").make_absolute(base_empty).to_string(), "/");
 
+        xlnt_assert_equals(xlnt::uri(".").make_absolute(base_empty).to_string(), "");
+        xlnt_assert_equals(xlnt::uri("..").make_absolute(base_empty).to_string(), "");
 
         xlnt_assert_equals(xlnt::uri("child").make_absolute(xlnt::uri("http://example.com")).to_string(), "http://example.com/child");
         xlnt_assert_equals(xlnt::uri("child").make_absolute(xlnt::uri("urn:base")).to_string(), "urn:child");
@@ -465,6 +469,12 @@ public:
 
         xlnt::uri ref2 = xlnt::uri("urn:target").make_reference(xlnt::uri("urn:base"));
         xlnt_assert_equals(ref2.to_string(), "target");
+
+        xlnt::uri ref_no_trailing = xlnt::uri("http://example.com/a/b/d.txt").make_reference(xlnt::uri("http://example.com/a/b/c"));
+        xlnt_assert_equals(ref_no_trailing.to_string(), "d.txt");
+
+        xlnt::uri ref_up = xlnt::uri("http://example.com/a/d.txt").make_reference(xlnt::uri("http://example.com/a/b/c/"));
+        xlnt_assert_equals(ref_up.to_string(), "../../d.txt");
     }
 };
 

